src/Interpreters/Cache/FileSegment.cpp

```diff
@@ -341,9 +334,40 @@ void FileSegment::setRemoteFileReader(RemoteFileReaderPtr remote_file_reader_)
 }
 
 void FileSegment::write(const char * from, size_t size, size_t offset)
+{
+    ProfileEventTimeIncrement<Microseconds> watch(ProfileEvents::FileSegmentWriteMicroseconds);
+    assertWriteAllowed(size, offset);
+    writeImpl(size, offset, [&]()
+    {
+        if (!cache_writer)
+            cache_writer = std::make_unique<WriteBufferFromFile>(getPath());
+
+        cache_writer->write(from, size);
+        cache_writer->next();
+    });
+}
+
+void FileSegment::write(WriteBufferFromFile & wb, size_t offset)
 {
     ProfileEventTimeIncrement<Microseconds> watch(ProfileEvents::FileSegmentWriteMicroseconds);
 
+    const size_t size = wb.offset();
+    assertWriteAllowed(size, offset);
+
+    const auto & current_filename = wb.getFileName();
+    const auto & expected_filename = getPath();
+    if (current_filename != expected_filename)
+    {
+        throw Exception(ErrorCodes::LOGICAL_ERROR,
+                        "Write buffer has unexpected file path: {}, expected: {}",
+                        current_filename, expected_filename);
+    }
+
+    writeImpl(size, offset, [&]() { wb.next(); }, &wb);
```

Comment:
Is it correct comment?

```suggestion
    /// Data should already be in buffer, so we only call `next`
    writeImpl(size, offset, [&]() { wb.next(); }, &wb);
```